### DISCUSSION TITLE
Using function sql, to cast type number to string, generate wrong query

### DISCUSSION THREAD

**Discussion by duyvinh229148** [mikro-orm/mikro-orm#6052] [General]
### Describe the bug

I have column 'shdon' of type number, I want to cast it to string so I can search it with partial match
This is filterQuery I'm using
```findQueryObject[sql`shdon::text`] = { $ilike: `%${soHoaDon}%` };```
I remember this used to work for me, when I use version range ^6.3.7
But today it throw this error:
```InvalidFieldNameException: select count(*) as "count" from "invoice"."hoa_don" as "h0" where "h0"."tai_khoan_hddt_id" = 48 and "h0"."sell_buy_type" = 'mua' and "h0"."tdlap" >= '2023-12-31T17:00:00.000Z' and "h0"."tdlap" <= '2024-09-21T07:00:00.000Z' and "h0"."[raw]: shdon::text (#1)" ilike '%4%' - column h0.[raw]: shdon::text (#1) does not exist```

Please help me with this, thanks

### Reproduction

I have never create a code sandbox with database before, so, sorry

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.3.10

### Node.js version

20

### Operating system

Linux - Fedora 39

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [x] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
Can't help much without seeing what you are doing. 

One common cause for this is reusing the same query for multiple find calls - the raw query fragments need to be unique, you can't reuse them.

```ts
// this will fail
const query = { [sql`foo`]: 1 };
await em.find(User, query);
await em.find(User, query);

// this will work
await em.find(User, { [sql`foo`]: 1 });
await em.find(User, { [sql`foo`]: 1 });
```

**Reply by BMO568** [reply]
Hi, I also faced a similar problem as you had described here.
it seems quite strange that it's required to recreate a query filter to run the same query. It would be useful if we could work with the same query filter as we do in the 5th version.
It could be convenient to store a query filter in some variable and use it in several "find" functions.

**Comment by baikinkirill** [comment]
Hi! Recently in our project, we also encountered this limitation when using fragments, and it created some problems for us. I made a small project to demonstrate our use case: https://github.com/baikinkirill/mikroorm-fragments-debug.

Currently, we solved this problem by generating the filter multiple times, however this is twice the load on our database and related microservices.

It would be very convenient if this limitation could be removed.

**Reply by B4nan** [reply]
that project is way too huge for me to easily see what you have problem with. what i mentioned above is still valid, you cant use a raw fragment instance multiple times, and that limitation wont be removed, since it would mean opening a security vulnerability. not sure how it would force you to do some work twice, the only thing that you need to do twice is instantiating the raw fragment (which has no real cost, its just about instantiating a class instance, no db involved in that), no reason to duplicate filters.

if your logic which prepares the condition is costly, just do that once, and only create the raw fragments multiple times.

**Reply by baikinkirill** [reply]
Caching as one of the options. However, I also noticed that, for example, in PostgreSQL the parameters in the fragment are escaped each time with every query, and this breaks the second and subsequent queries. Can I in this case form a RawQueryFragment myself, preprocess it once, and insert the result into the filter?

```js
const fragment = new RawQueryFragment(`SELECT 1 WHERE title = ${condition}`);

const filter = {
    [fragment.toPrepared()]: []
};

const firstResult = em.find(Entity, filter);
const secondResult = em.find(Entity, filter);
```

**Reply by B4nan** [reply]
idk what you mean

**Reply by Ruby184** [reply]
> it would mean opening a security vulnerability.

What exactly do you mean by that, @B4nan? If you're referring to memory leaks, that can be addressed by avoiding strong references to `RawQueryFragment`. Instead, we could use a `WeakMap` with `Symbol` keys, allowing the objects to be garbage collected once they're no longer referenced. I’d be happy to open an RFC explaining the solution if you're interested.

**Reply by B4nan** [reply]
The issue here is that you can use the fragment as a key, which stringifies it, and we need a map of those to be able to convert back. The object won't be referenced anywhere once it's serialized, we can't use a weak map for it. The vulnerability part is about not having this cache, in other words, always deserializing a stringified raw fragment - you could forge a query that would have a raw part from outside this way. 

edit: or you mean you'd create a new symbol for each raw instance and use that as a key? if that works, PR welcome, I would love to get rid of the cache

**Reply by Ruby184** [reply]
> edit: or you mean you'd create a new symbol for each raw instance and use that as a key? if that works, PR welcome, I would love to get rid of the cache

Yes, the idea is to use symbols instead of strings to maintain a weak reference to the fragment. When a `RawQueryFragment` is used as a key in an object query, a unique symbol is created for each usage to serve as a reference. This allows us to store the fragment in a weak reference using:

```ts
static #rawQueryReferences = new WeakMap<symbol, RawQueryFragment>();
```
This way, when the symbol is no longer referenced (i.e., no object uses it as a key), the associated RawQueryFragment can be automatically garbage collected.

I’m currently working on a PR and will have a proof of concept ready shortly.

**Reply by Ruby184** [reply]
@B4nan There's a minor issue with the ordering of keys in an object when combining string and symbol keys (e.g., raw fragments). According to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Reflect/ownKeys#return_value), when retrieving keys via Reflect.ownKeys(), the order is:

1. Non-negative integer indices in ascending order (as strings),
2. Other string keys in the order of property creation,
3. Symbol keys in the order of property creation.

This causes an issue when using orderBy with a mix of raw fragments (symbols) and standard string properties. For example:

```ts
await orm.em.findByCursor(Tag, {}, {
  populate: ['job'],
  first: 3,
  orderBy: {
    [raw(alias => `${alias}.created`)]: 'desc',
    job: {
      [raw(alias => `${alias}.DateCompleted`)]: 'desc',
    },
  },
});
```
The expected order is: `t0.created desc, j1.DateCompleted desc`.
But due to the symbol key from raw fragment, the actual order becomes: `j1.DateCompleted desc, t0.created desc`.

Since ordering is critical in this context, I propose disallowing the combination of raw fragments and string keys in a single `orderBy` object. Instead, we should require the use of an array of order definitions. This would be a breaking change, so I suggest targeting the v7 branch.

Under the new rule, the example above would throw an error. The correct usage would be:

```ts
await orm.em.findByCursor(Tag, {}, {
  populate: ['job'],
  first: 3,
  orderBy: [
    { [raw(alias => `${alias}.created`)]: 'desc' },
    { job: { [raw(alias => `${alias}.DateCompleted`)]: 'desc' } },
  ],
});
```

What do you think?

**Reply by B4nan** [reply]
Yeah, makes sense, let's target v7 then.

**Reply by B4nan** [reply]
@Ruby184 could you PR those changes to v7? Can be just a draft, I'll be happy to finish it.

**Reply by Ruby184** [reply]
@B4nan Yeah, I need to merge the changes from v7 and pick it up from where I left off, but I’ll open at least a draft so you can see the concept.

**Reply by B4nan** [reply]
Friendly ping, still waiting for the draft PR, I guess I know how to do this myself, but still, I would appreciate a look at your version before I dive into it (also I would very much like to reuse any tests you might have written already).

I'd like to merge v7 to master (and ship first RC) in a month or so, and I'd like this to be included in there.

**Reply by Ruby184** [reply]
Thanks for the ping! I’ve been a bit busy, but I should be able to finish this over the weekend. I’ll share the draft PR as soon as it’s ready so you can review it and reuse what I have (will need a few things to clean up and refactor). Appreciate the patience.

**Reply by B4nan** [reply]
I gave it a try, and it feels like we might need to replace _a lot_ of places that use `Object.keys` with `Reflect.ownKeys`. When I was trying to make some tests work, I was just replacing all occurrences of it without too much thinking, but it didn't feel right 🥲 But finding the code paths that needed changing was quite exhausting.

Do I get it right that we can get rid of all the hacks around the cleanup, return the same instance when cloning, and skip the ALS trick for `findAndCount`?

**Reply by Ruby184** [reply]
@B4nan Yes, I’ve spent a lot of time refactoring every use case where object queries are handled and trying to correctly map the conditions for raw fragments. All hacks requiring cleanup and cloning can be dropped, as well as the ALS storage. I’ve been a bit busy with personal life and work, but I’ll try to put this together for you today. I just need to fix a few things and write down some TODO questions I encountered so you can resolve them.
